lib/token-store.js

```diff
@@ -8,196 +8,1110 @@
+    /** @inheritdoc */
+    moveNext() {
+        const retv = this.cursor.moveNext();
+
+        this.current = this.cursor.current;
+
+        return retv;
+    }
+
+    /** @inheritdoc */
+    movePrev() {
+        const retv = this.cursor.movePrev();
+
+        this.current = this.cursor.current;
+
+        return retv;
+    }
+}
+
+/**
+ * The decorative cursor which ignores specified tokens.
+ * @private
+ */
+class FilterCursor extends DecorativeCursor {
 
     /**
-     * Gets the token that follows a given node or token in the token stream.
-     * @param {(ASTNode|Token)} node The AST node or token.
-     * @param {int} [skip=0] A number of tokens to skip after the given node or
-     *     token.
-     * @returns {Token} An object representing the token.
+     * Initializes this cursor.
+     * @param {Cursor} cursor - The cursor to be decorated.
+     * @param {Function} predicate - The predicate function to decide tokens this cursor iterates.
      */
-    api.getTokenAfter = function(node, skip) {
-        return tokens[lastTokenIndex(node) + (skip || 0) + 1];
-    };
+    constructor(cursor, predicate) {
+        super(cursor);
+        this.predicate = predicate;
+    }
+
+    /** @inheritdoc */
+    moveNext() {
+        while (super.moveNext()) {
+            if (this.predicate(this.current)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /** @inheritdoc */
+    movePrev() {
+        while (super.movePrev()) {
+            if (this.predicate(this.current)) {
+                return true;
+            }
+        }
+        return false;
+    }
+}
+
+/**
+ * The decorative cursor which ignores the first few tokens.
+ * @private
+ */
+class SkipCursor extends DecorativeCursor {
 
     /**
-     * Gets all tokens that are related to the given node.
-     * @param {ASTNode} node The AST node.
-     * @param {int} [beforeCount=0] The number of tokens before the node to retrieve.
-     * @param {int} [afterCount=0] The number of tokens after the node to retrieve.
-     * @returns {Token[]} Array of objects representing tokens.
+     * Initializes this cursor.
+     * @param {Cursor} cursor - The cursor to be decorated.
+     * @param {number} count - The count of tokens this cursor skips.
      */
-    api.getTokens = function(node, beforeCount, afterCount) {
-        return get(
-            starts[node.range[0]] - (beforeCount || 0),
-            lastTokenIndex(node) + (afterCount || 0) + 1
-        );
-    };
+    constructor(cursor, count) {
+        super(cursor);
+        this.count = count;
+    }
+
+    /** @inheritdoc */
+    moveNext() {
+        while (this.count > 0) {
+            this.count -= 1;
+            if (!super.moveNext()) {
+                return false;
+            }
+        }
+        return super.moveNext();
+    }
+
+    /** @inheritdoc */
+    movePrev() {
+        while (this.count > 0) {
+            this.count -= 1;
+            if (!super.movePrev()) {
```

Comment:
It looks like the `skip` behavior depends on whether `moveNext` or `movePrev` is called first. If `moveNext` is called first, then the *first* `count` tokens are skipped, but if `movePrev` is called first, then the *last* `count` tokens are skipped.

I'm not sure if there's an easy workaround for this, though.